### ISSUE TITLE
[clang][analyzer] security.ArrayBound false positive after popcount loop

### ISSUE THREAD

**Issue by bozicrHT** [llvm/llvm-project#224227]
**Reproducer:**
```c
#include <stdlib.h>

void test(unsigned bits)
{
    unsigned copy = bits;
    unsigned count = 0;
    unsigned index = 0;

    for (; copy; copy &= copy - 1)
        ++count;

    char *output = malloc(count);
    if (!output)
        return;

    if (bits & 1)
        output[index++] = 0;

    if (bits & 2)
        output[index++] = 0; /* False positive. */

    free(output);
}
```
The analyzer reports the second write as past a one-byte allocation. That path is impossible:
- A one-byte allocation means the popcount loop determined `count == 1`.
- Reaching both writes means both `bits & 1` and `bits & 2` are nonzero.
- That necessarily means `popcount(bits) >= 2`.

**Independent demonstration:**
```c
void test(unsigned bits)
{
    unsigned copy = bits;
    unsigned count = 0;

    for (; copy; copy &= copy - 1)
        ++count;

    if ((bits & 1) && (bits & 2) && count < 2)
        clang_analyzer_warnIfReached(); /* Incorrectly reports REACHABLE. */
}
```

Compiler explorer link can be found [here.](https://compiler-explorer.com/z/zz5adTWs9)

**Comment by bozicrHT** [comment]
I don't think this is primarily a `security.ArrayBound` problem. That warning appears to be the manifestation of an infeasible state that survived constraint solving. <https://compiler-explorer.com/z/ddh8b4WYa>

Consider this reduced example, which contains no allocation or bounds access:
```c
void clang_analyzer_warnIfReached(void);
void clang_analyzer_printState();
void clang_analyzer_eval(int);

void test(unsigned bits) {
  if ((bits & (bits - 1U)) == 0 &&
      (bits & 1U) != 0 &&
      (bits & 2U) != 0)
    clang_analyzer_printState();

    clang_analyzer_eval((bits & (bits-1U)) == 0);
    clang_analyzer_eval((bits & 1U) != 0);
    clang_analyzer_eval((bits & 2U) != 0);

    clang_analyzer_warnIfReached(); // Reports REACHABLE.
}
```
The relevant part of the state is:
```json
{ "symbol": "(reg_$1<unsigned int bits>) & ((reg_$1<unsigned int bits>) - 1U)", "range": "{ [0, 0] }" },
{ "symbol": "(reg_$1<unsigned int bits>) & 1U", "range": "{ [1, 1] }" },
{ "symbol": "(reg_$1<unsigned int bits>) & 2U", "range": "{ [1, 2] }" }
```
The condition is unsatisfiable. The underlying issue is likely in constraint handling for bitwise expressions.

**Question:** The `[0,0]` range means that the result of `bits & (bits - 1U)` is zero, not that `bits` itself is zero. Does the constraint manager understand that this means `bits` must be either zero or a power of two? If so, why is that information not printed?

**Comment by daniel-petrovic** [comment]
is this reproducable with z3 constraints ? 

AFAIK z3 smt solver is using bit-vector theory instead of range-based reasoning

enabling z3 requires building clang with -DLLVM_ENABLE_Z3_SOLVER=ON and passing via -Xanalyzer crosscheck-with-z3=true  .. not sure if there is appropriate clang build with z3 enabled on compiler explorer

**EDIT**: just verified on my machine: z3 eliminates false positive:

```yaml
./bin/clang --analyze \
  -Xanalyzer -analyzer-checker=security.ArrayBound \
  -Xanalyzer -analyzer-checker=debug.ExprInspection \
  -Xanalyzer -analyzer-config -Xanalyzer crosscheck-with-z3=true \
  -Xanalyzer -analyzer-output=text \
  /tmp/test.cpp
"program_state": {
  "store": { "pointer": "0x4a669888", "items": [
    { "cluster": "SymRegion{conj_$0{int &, LC1, S762, #0}}", "pointer": "0x4a65bc58", "items": [
      { "kind": "Direct", "offset": 0, "value": "0 S32b" }
    ]},
    { "cluster": "count", "pointer": "0x4a65ca90", "items": [
      { "kind": "Direct", "offset": 0, "value": "1 U32b" }
    ]}
  ]},
  "environment": { "pointer": "0x4a64a310", "items": [
    { "lctx_id": 1, "location_context": "#0 Call", "calling": "test", "location": null, "items": [
      { "stmt_id": 946, "kind": "ImplicitCastExpr", "pretty": "clang_analyzer_printState", "value": "&code{clang_analyzer_printState}" }
    ]}
  ]},
  "constraints": [
    { "symbol": "(reg_$1<unsigned int bits>) & ((reg_$1<unsigned int bits>) - 1U)", "range": "{ [0, 0] }" },
    { "symbol": "(reg_$1<unsigned int bits>) & 1U", "range": "{ [1, 1] }" },
    { "symbol": "(reg_$1<unsigned int bits>) & 2U", "range": "{ [1, 2] }" },
    { "symbol": "reg_$1<unsigned int bits>", "range": "{ [1, 4294967295] }" }
  ],
  "equivalence_classes": null,
  "disequality_info": null,
  "dynamic_types": null,
  "dynamic_casts": null,
  "checker_messages": null
}
`

**Comment by bozicrHT** [comment]
You're right! It **does eliminate** the false positive. However, it still reports the code as reachable for the reduced example:
```c
void test(unsigned bits) {
  if ((bits & (bits - 1U)) == 0 &&
      (bits & 1U) != 0 &&
      (bits & 2U) != 0)
    clang_analyzer_printState();

    clang_analyzer_eval((bits & (bits - 1U)) == 0);
    clang_analyzer_eval((bits & 1U) != 0);
    clang_analyzer_eval((bits & 2U) != 0);

    clang_analyzer_warnIfReached(); // Reports REACHABLE.
}
```
